gke-deploy/core/gcs/gcs_test.go

```diff
@@ -0,0 +1,163 @@
+
+	expandedGcsDst     = "gs://project/path/expanded/aggregated-resources.yaml"
+	bucketNotFoundDst  = "gs://project/not_found_bucket/expanded/aggregated-resources.yaml"
+	accessDeniedBucket = "gs://project/access_denied_bucket/expanded/aggregated-resources.yaml"
+
+	errMsg            = "copy file failed"
+	errDenied         = "error is AccessDeniedException: 403"
+	errFileNotFound   = "error is CommandException: No URLs matched"
+	errBucketNotFound = "error is NotFoundException: 404 The destination bucket does not exist"
+)
+
+func buildTestGCS(t *testing.T) *GCS {
+	t.Helper()
+
+	s := &testservices.TestGcsService{CopyResponse: map[string]func(src, dst string, recursive bool) error{
+		singleFile: func(src, dst string, recursive bool) error { return nil },
+		directory:  func(src, dst string, recursive bool) error { return nil },
+		nestedDir:  func(src, dst string, recursive bool) error { return nil },
+		slowReadFile: func(src, dst string, recursive bool) error {
+			time.Sleep(3 * time.Second)
+			return nil
+		},
+		errorFile:         func(src, dst string, recursive bool) error { return errors.New(errMsg) },
+		accessDeniedFile:  func(src, dst string, recursive bool) error { return errors.New(errDenied) },
+		notFoundFile:      func(src, dst string, recursive bool) error { return errors.New(errFileNotFound) },
+		expandedK8sConfig: func(src, dst string, recursive bool) error { return nil },
+		slowUploadConfig: func(src, dst string, recursive bool) error {
+			time.Sleep(3 * time.Second)
+			return nil
+		},
+		accessDeniedConfig:   func(src, dst string, recursive bool) error { return errors.New(errDenied) },
+		bucketNotFoundConfig: func(src, dst string, recursive bool) error { return errors.New(errBucketNotFound) },
+	},
+	}
+
+	return &GCS{
+		Timeout:    timeoutGCS,
+		Retries:    retries,
+		Delay:      delay,
+		GcsService: s,
+	}
+
+}
+
+func TestFetch(t *testing.T) {
+
+	tests := []struct {
+		name      string
+		src       string
+		dst       string
+		recursive bool
+		ok        bool
+		status    string
+	}{
+		{"download single file", singleFile, dstDir, false, true, "pass"},
+		{"download files from a directory", directory, dstDir, false, true, "pass"},
+		{"download files from a directory with nested folders", nestedDir, dstDir, true, true, "pass"},
+		{"download files failed on timeout", slowReadFile, dstDir, false, false, "fail"},
+		{"download files with failure", errorFile, dstDir, false, false, "fail"},
+		{"download files with access denied failure", accessDeniedFile, dstDir, false, false, "fail"},
+		{"download files with not found failure", notFoundFile, dstDir, false, false, "fail"},
+	}
+
+	gcs := buildTestGCS(t)
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+
+			err := gcs.Download(context.Background(), tc.src, tc.dst, tc.recursive)
+			if (err == nil) != tc.ok {
+				t.Errorf("Test should %s but not", tc.status)
```

Comment:
This codebase tries to follow a pattern of describing what args were passed in for the failing case: e.g., https://github.com/GoogleCloudPlatform/cloud-builders/blob/master/gke-deploy/core/cluster/cluster_test.go#L27

Reply:
Add a test for errors.